src/main/java/io/cdap/plugin/gcp/gcs/sink/ForwardingOutputCommitter.java

```diff
@@ -0,0 +1,110 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.gcs.sink;
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import io.cdap.cdap.api.exception.ErrorType;
+import io.cdap.plugin.gcp.common.GCPUtils;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import java.io.IOException;
+
+/**
+ * ForwardingOutputCommitter which delegates all operations to another OutputCommitter.
+ * <p>
+ * This is used to wrap the OutputCommitter of the delegate format and
+ * throw {@link io.cdap.cdap.api.exception.ProgramFailureException} from IOException.
+ */
+public class ForwardingOutputCommitter extends OutputCommitter {
+  private final OutputCommitter delegate;
+
+  public ForwardingOutputCommitter (OutputCommitter delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public void setupJob (JobContext jobContext) throws IOException {
+    try {
+      delegate.setupJob(jobContext);
+    } catch (IOException e) {
+      getException(e);
+    }
+  }
+
+  @Override
+  public void setupTask (TaskAttemptContext taskAttemptContext) throws IOException {
+    try {
+      delegate.setupTask(taskAttemptContext);
+    } catch (IOException e) {
+      getException(e);
+    }
+  }
+
+  @Override
+  public boolean needsTaskCommit (TaskAttemptContext taskAttemptContext) throws IOException {
+    try {
+      return delegate.needsTaskCommit(taskAttemptContext);
+    } catch (IOException e) {
+      getException(e);
+    }
+    return false;
+  }
+
+  @Override
+  public void commitTask (TaskAttemptContext taskAttemptContext) throws IOException {
+    try {
+      delegate.commitTask(taskAttemptContext);
+    } catch (IOException e) {
+      getException(e);
+    }
+  }
+
+  @Override
+  public void abortTask (TaskAttemptContext taskAttemptContext) throws IOException {
+    try {
+      delegate.abortTask(taskAttemptContext);
+    } catch (IOException e) {
+      getException(e);
+    }
+  }
+
+  @SuppressWarnings("rawtypes")
+  public void addGCSOutputCommitterFromOutputFormat(OutputFormat outputFormat,
+      String tableName) throws IOException, InterruptedException {
+    if (delegate instanceof DelegatingGCSOutputCommitter) {
+      try {
+        ((DelegatingGCSOutputCommitter) delegate).addGCSOutputCommitterFromOutputFormat(
+            outputFormat, tableName);
+      } catch (IOException e) {
+        getException(e);
+      }
+      return;
+    }
+    throw GCPUtils.getProgramFailureException(null,
+        String.format("Operation is not supported in the output committer: %s",
+            delegate.getClass().getName()), null, ErrorType.SYSTEM, false);
+  }
+
+  private void getException(IOException e) throws IOException {
+    if (e instanceof GoogleJsonResponseException) {
```

Comment:
I would recommend to have this logic in the utils class. This would allow to handle more exception types later
Also by looking at getProgramFailureException, it does not seem that you need `GoogleJsonResponseException`, `HttpResponseException` should give you all data you need.

Another consideration is to have throw in the calling code, similar to guava `Throwables.propagate`. it would make it easier for static analysis (and viewer) to know that it's a throwing statement

Reply:
done.